File: leetcode/januarychallenge/mergeTwoLists.go
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeTwoLists(n1 *ListNode, n2 *ListNode) *ListNode {
	iterator2 := n1
	if n2 == nil {
		return n1
	}
	if n1 == nil {
		return n2
	}
	for iterator2 != nil {
		n2 = SortListInsert(n2, iterator2.Val)
		iterator2 = iterator2.Next
	}
	return n2
}

func SortListInsert(l *ListNode, data_ref int) *ListNode {
	n := &ListNode{Val: data_ref}
	if l == nil {
		return n
	}
	if n.Val < l.Val {
		n.Next = l
		return n
	}
	iterator := l
	for iterator.Next != nil && n.Val > iterator.Next.Val {
		iterator = iterator.Next
	}
	n.Next = iterator.Next
	iterator.Next = n
	return l
}

// func mergeTwoLists(l1 *ListNode, l2 *ListNode) *ListNode {
//     if l1 == nil{
//         return l2
//     }
//     if l2 == nil{
//         return l1
//     }
//     if l1.Val < l2.Val {
//         l1.Next = mergeTwoLists(l1.Next, l2)
//         return l1
//     }else {
//         l2.Next = mergeTwoLists(l2.Next, l1)
//         return l2
//     }
// }

File: algo/mergeTwoLists.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode() {}
 *     ListNode(int val) { this.val = val; }
 *     ListNode(int val, ListNode next) { this.val = val; this.next = next; }
 * }
 */


class MergeTwoLists {
    // Time complexity : O(n+m)
    // Space complexity : O(1)
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        
        ListNode res = new ListNode();
        ListNode temp = res;
        
        while (l1 != null && l2 != null) {
            if (l1.val < l2.val) {
                temp.next = l1;
                l1 = l1.next;
            } else {
                temp.next = l2;
                l2 = l2.next;
            }
            temp = temp.next;
        }
        
        if (l1 != null) {
            temp.next = l1; 
        } else if (l2 != null) {
            temp.next = l2;
        }
        
        return res.next;   
    }

    // Recursion
    // Time complexity : O(n+m)
    // Space complexity : O(n+m)
    public ListNode mergeTwoLists(ListNode l1, ListNode l2) {
        if (l1 == null) {
            return l2;
        }
        else if (l2 == null) {
            return l1;
        }
        else if (l1.val < l2.val) {
            l1.next = mergeTwoLists(l1.next, l2);
            return l1;
        }
        else {
            l2.next = mergeTwoLists(l1, l2.next);
            return l2;
        }

    }
}